### ISSUE TITLE
941160 reacts to img

### ISSUE THREAD

**Issue by Simbiat** [coreruleset/coreruleset#3770]
### Description
I am using Coraza as module in Caddy with below directives:
```
Include @coraza.conf-recommended
		Include @crs-setup.conf.example
		Include @owasp_crs/*.conf
		SecRuleEngine On
```
When submitting a POST request with `img` tag request gets 403 with reference to 941160 ID, which does not seem to even be related to `img`. Tried changing different things in `img`, but only complete removal of it or leaving it as `<img>` solves the problem. I get that it should be reacting to `img src`, meaning, that I probably need to somehow disable or adjust the rule (can I somehow allow src with relative paths only?), but even if I leave `<img >` - it still matches, although from XSS attack this looks safe (and pointless).

### How to reproduce the misbehavior (-> curl call)
For "baseline":
```
curl -X POST https://sandbox.coreruleset.org/ \
-H "Content-Type: application/x-www-form-urlencoded" \
-d 'test=<p><img class="w50pc middle block galleryZoom" src="/assets/images/uploaded/7c/e9/f4/7ce9f467294f6e47f19224acd6d35a63f49079448118d9cc5fe6a326221987e030afd319c17d5f571b3b87adf27024e7983a9709e873ee52b2f24b7d7a9111e9.webp" alt="Dream apartment plan"></p>'
```
Something that actually seems faulty to me:
```
curl -X POST https://sandbox.coreruleset.org/ \
-H "Content-Type: application/x-www-form-urlencoded" \
-d 'test=<p><img ></p>'
```

### Logs
Curls provided

### Your Environment

Using Coraza https://github.com/corazawaf/coraza-caddy, not sure how to see which version is used there

### Confirmation

[X ] I have removed any personal data (email addresses, IP addresses,
    passwords, domain names) from any logs posted.

**Comment by airween** [comment]
Hi @Simbiat,

thanks for reporting.

Please take a look at the rule [941160](https://github.com/coreruleset/coreruleset/blob/main/rules/REQUEST-941-APPLICATION-ATTACK-XSS.conf#L189-L219) and its comment. There you can see how the regex is built.

[Here](https://github.com/coreruleset/coreruleset/blob/main/regex-assembly/941160.ra) you can see the parts of the regex what is mentioned in the comment.

The data file has many blocks, but (I think) these two parts which produces that the `<img ` pattern matches with that regex:

* [this](https://github.com/coreruleset/coreruleset/blob/main/regex-assembly/941160.ra#L7) line adds the `<` character
* [this](https://github.com/coreruleset/coreruleset/blob/main/regex-assembly/941160.ra#L27) matches with the `img ` (with the space at the end)

Please take a look: https://regex101.com/r/Dhj15K/1 with pattern `img ` (with the space at the end).

And I think that's intended, not accident (so may be it's not FP).

I suggest you to create an exclusion to avoid this behavior.

**Comment by Simbiat** [comment]
I have checked the regex before creating the issue, and I do understand why it matches in my case, but I still think this may be a bug, because it matches even an `img` without `src` or any of the event attributes. Yes, such HTML is unlikely to happen, but you can theoretically have legitimate `img`, that are then populated through JS based on some logic. That, in turn, may become an XSS vulnerability, but that would be beyond the scope of this particular rule.

**Comment by airween** [comment]
Okay, we will take a look it on our next [monthly chat](https://github.com/coreruleset/coreruleset/issues/3753).

Thanks again.

**Comment by Simbiat** [comment]
Speaking of exclusion for this, I've tried various things, now I've come to this:
```
# Allow HTML when submitting new posts
# Phase 1: Identify and mark the request for exclusion
SecRule REQUEST_METHOD "POST" "phase:1,nolog,pass,id:10000"
SecRule REQUEST_URI "^/api/" "phase:1,nolog,pass,chain"
SecRule ARGS_NAMES "@rx ^postForm\[text\]$" "id:10001,phase:1,pass,nolog,ctl:requestBodyAccess=On,ctl:ruleRemoveById=941160"
# Phase 2: Apply the exclusion based on the marking from phase 1
SecRule REQUEST_METHOD "POST" "phase:2,nolog,pass,id:10002"
SecRule REQUEST_URI "^/api/" "phase:2,nolog,pass,chain"
SecRule ARGS_NAMES "@rx ^postForm\[text\]$" "id:10003,phase:2,pass,nolog,ctl:ruleRemoveById=941160"
Include @coraza.conf-recommended
Include @crs-setup.conf.example
Include @owasp_crs/*.conf
SecRuleEngine On
```
It still blocks the requests, though. I also tried putting `SecRule` after includes, too, still no luck. Although I have little idea of what I am doing. What I want is to ignore 941160 rule for `postForm\[text\]` field in POST request to specific URI. What am I doing wrong?

**Comment by airween** [comment]
> Speaking of exclusion for this, I've tried various things, now I've come to this:
> 
> ```
> # Allow HTML when submitting new posts
> # Phase 1: Identify and mark the request for exclusion
> SecRule REQUEST_METHOD "POST" "phase:1,nolog,pass,id:10000"
> SecRule REQUEST_URI "^/api/" "phase:1,nolog,pass,chain"
> SecRule ARGS_NAMES "@rx ^postForm\[text\]$" "id:10001,phase:1,pass,nolog,ctl:requestBodyAccess=On,ctl:ruleRemoveById=941160"
> # Phase 2: Apply the exclusion based on the marking from phase 1
> SecRule REQUEST_METHOD "POST" "phase:2,nolog,pass,id:10002"
> SecRule REQUEST_URI "^/api/" "phase:2,nolog,pass,chain"
> SecRule ARGS_NAMES "@rx ^postForm\[text\]$" "id:10003,phase:2,pass,nolog,ctl:ruleRemoveById=941160"
> Include @coraza.conf-recommended
> Include @crs-setup.conf.example
> Include @owasp_crs/*.conf
> SecRuleEngine On
> ```
> 
> It still blocks the requests, though. I also tried putting `SecRule` after includes, too, still no luck. Although I have little idea of what I am doing. What I want is to ignore 941160 rule for `postForm\[text\]` field in POST request to specific URI. What am I doing wrong?

I try to understand your rules above, but why is there one with `id:10000`?
And the next rule does not have any `id`.

Same as with the `id:10002` and the next.

Perhaps you miss the `chain`?

If I were you I would take a look it at again... May be this is what you would have wanted:

```
SecRule REQUEST_METHOD "@strEq POST" \
    "id:10000,\
    phase:1,\
    nolog,\
    pass,\
    chain"
    SecRule REQUEST_URI "@rx ^/api/" \
        "chain"
            SecRule ARGS_NAMES "@rx ^postForm\[text\]$"  \
                "id:10001,\
                ctl:requestBodyAccess=On,\
                ctl:ruleRemoveById=941160"
```

(but I don't see the reason the `ctl:requestBodyAccess=On`).

Based on this, you should take a review with the other (chained) rule.

Please pay more attention of these:
* a chained rule must have only one `id`
* only one disruptive action (`pass`)
* use explicit operators (see ModSecurity's issue [#3081](https://github.com/owasp-modsecurity/ModSecurity/issues/3081))

**Comment by Simbiat** [comment]
As I've mentioned I had no idea what I was doing (never had a need to customize ModSecujrity rules before), so I relied on GPT. Digging a bit more into documentation I've found a suitable way to do this, which I shared in the previous comment. Since it was missed, I would like to highlight again, that `img` was also triggering 942100 rule, after 941160 was skipped for it, so this may need to be included in the planned discussion, too.

**Comment by theseion** [comment]
I agree that the rule should not match `<img `. I'm looking into it, but it could take a while.

**Comment by theseion** [comment]
I've done some digging. The rule is ancient. It was plucked from the noscript browser extension, somewhere around version 1 or 2 (source code available here: https://noscript.net/getit/). The names `js-prop-prefix` and `js-payloads` in the assembly file are wrong. Here's a corrected version with comments:

```
##!> assemble
  ##! HTML tag opening bracket
  <
  ##!=>
  ##! Quotes, punctuation, brackets, etc.
  [^\w<>]*
  ##!=>
  ##! More or less `\w`, followed by `:`, which would match `xml:base`, together with the
  ##! list defined later.
  (?:[^<>\"'\s]*:)?
  ##!=>
  ##! Quotes, punctuation, brackets, etc.
  [^\w<>]*
  ##!=>
  ##! Quotes, punctuation, brackets, etc., angle brackets allowed
  \W*?
  ##!=< html-tag-prefix
##!<
```

What this is supposed to match, is the start of one of a list of tags considered dangerous _w.r.t. noscript_. This list includes `img`.  

Now, even though that's probably not exactly what makes the most sense for CRS, the rule does exactly what it advertises, it detects HTML injection. Whether he HTML is valid or not doesn't really make a difference.

We'll have to discuss in the team what to do about the rule. In general, however, it is quite normal for CRS to block on HTML in the request payload.

**Comment by Simbiat** [comment]
I agree, that CRS should block HTML by default, but considering WYSIWIG editors exist (at least), in some cases you will need to allow HTML through and sanitize it with some other means rather than block it entirely. Perhaps it would make sense to move this to a separate rule? I mean 941160 rule seems to be checking quite a few things at once

**Comment by theseion** [comment]
> Perhaps it would make sense to move this to a separate rule? I mean 941160 rule seems to be checking quite a few things at once

Yes, possibly.

**Comment by theseion** [comment]
We've discussed this issue in the team. We understand that this rule is not ideal, however, it's aggressiveness also makes it good at its job, which is to detect _any_ HTML. We do not have the resources to make adjustments to this rule at the moment, as we have to prioritise other work, though we do welcome pull requests, if you would like to help out.

Our usual advice in these situations is for people to create rule exclusions (which you are familiar with, I believe), or to adjust the behaviour of their application, if they can.

I've opened a PR that improves the documentation of the rule, so at least we now know what it actually does :)
